omod/src/test/java/org/openmrs/module/owa/web/controller/AddAppControllerTest.java

```diff
@@ -46,11 +52,34 @@ public void restoreOriginalMessageSourceService() {
 	 * Test of upload method, of class AddAppController.
 	 */
 	@Test
-	public void testUpload() throws Exception {
-		System.out.println("upload");
+	public void testUploadofNotaZipArchieve() throws Exception {
+		HttpServletRequest request = new MockHttpServletRequest(new MockServletContext(), "POST", "/module/owa/addApp.htm");
+		URL url = OpenmrsClassLoader.getInstance().getResource("testing");
+		MockMultipartFile multifile = new MockMultipartFile("testFile", "testing", null, new FileInputStream(url.getFile()));
+		AddAppController controller = (AddAppController) applicationContext.getBean("addAppController");
+		controller.upload(multifile, request);
+		Assert.assertEquals(request.getSession().getAttribute("errormessage"), "owa.not_a_zip");
+	}
+	
+	@Test
+	public void testUploadofEmptyFile() throws Exception {
 		HttpServletRequest request = new MockHttpServletRequest(new MockServletContext(), "POST", "/module/owa/addApp.htm");
-		AddAppController instance = new AddAppController();
-		String expResult = "";
+		URL url = OpenmrsClassLoader.getInstance().getResource("file");
+		MockMultipartFile multifile = new MockMultipartFile("testFile", "file", null, new FileInputStream(url.getFile()));
+		AddAppController controller = (AddAppController) applicationContext.getBean("addAppController");
+		controller.upload(multifile, request);
+		Assert.assertEquals(request.getSession().getAttribute("errormessage"), "owa.blank_zip");
 	}
 	
+	@Test
+	public void testUploadofZipFilewithoutManifest() throws Exception {
+		HttpServletRequest request = new MockHttpServletRequest(new MockServletContext(), "POST", "/module/owa/addApp.htm");
+		URL url = OpenmrsClassLoader.getInstance().getResource("Zipwithoutmanifest.zip");
+		MockMultipartFile multifile = new MockMultipartFile("testFile", "Zipwithoutmanifest.zip", null,
+		        new FileInputStream(url.getFile()));
+		AddAppController controller = (AddAppController) applicationContext.getBean("addAppController");
+		controller.upload(multifile, request);
```

Comment:
@sunbiz  Irrespective of the zip this line is throwing java.util.zip.ZipException: error in opening zip file
	at java.util.zip.ZipFile.open(Native Method)
	at java.util.zip.ZipFile.<init>(ZipFile.java:220)
	at java.util.zip.ZipFile.<init>(ZipFile.java:150)
	at java.util.zip.ZipFile.<init>(ZipFile.java:164)
	at org.openmrs.module.owa.web.controller.AddAppController.upload(AddAppController.java:81)
	I am unable to figure out the reason 
But the same works fine when tested manually by uploading in UI Screen